lib/ecto/repo/queryable.ex

```diff
@@ -427,6 +427,15 @@ defmodule Ecto.Repo.Queryable do
     end)
   end
 
+  defp process_merge(row, {:source, {source, schema}, prefix, types}, _from, adapter) do
+    struct = Ecto.Schema.Loader.load_struct(schema, prefix, source)
+    struct_load!(types, row, [], false, struct, adapter)
```

Comment:
The `false` here is the key change. This makes it so that `struct_load!` does not convert the map to `nil` if all the values are `nil`. Normally this is called with `true`.

The alternative was to add a new parameter to all the `process` calls named `keep_nil?` or something like that. I chose this way because we only care about it in this one circumstance. So it keeps the noise to a minimum.